File: leetcode/num46.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(nums, [], res)
        return res
        
    def dfs(self, nums, path, res):
        if not nums:
            res.append(path)
        for i in range(len(nums)):
            self.dfs(nums[:i]+nums[i+1:], path+[nums[i]], res)

if __name__ == "__main__":
    solution = Solution()
    nums = [1, 2, 3]
    res = solution.permute(nums)
    print(res)

File: leetcode/num46.java
package leetcode;
import java.util.ArrayList;
import java.util.List;
public class num46{
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        if(nums == null || nums.length == 0) return result;
        backtrack(result, new ArrayList<>(), nums);
        return result;
    }
    public void backtrack(List<List<Integer>> res, List<Integer> temp, int[] nums){
        if(temp.size() == nums.length){
            res.add(new ArrayList<>(temp));
        }else{
            for(int i=0; i<=temp.size(); i++){
                temp.add(i, nums[temp.size()]);
                backtrack(res, temp, nums);
                temp.remove(i);
            }
        }
    }
    public static void main(String[] args) {
        int[] nums = {1, 2, 3};
        num46 solution = new num46();
        List<List<Integer>> result = solution.permute(nums);
        for(List<Integer> list:result){
            for(Integer item:list){
                System.out.print(item + "\t");
            }
            System.out.println();
        }
    }
}